File: skills/code-architect/references/code-examples/ruby/template_method.rb
# frozen_string_literal: true

# Template Method — algorithm skeleton in a base class with overridable steps.

# Base class fixes the algorithm order; subclasses fill in the steps.
class ImportPipeline
  # Template method: the invariant skeleton.
  def run(source)
    raw = read(source)
    rows = parse(raw)
    rows = filter(rows) # hook with a default
    store(rows)
  end

  private

  def read(_source) = raise NotImplementedError
  def parse(_raw) = raise NotImplementedError

  # Hook: optional override, sensible default.
  def filter(rows) = rows

  def store(rows) = "stored #{rows.size} rows: #{rows.inspect}"
end

class CsvImport < ImportPipeline
  private

  def read(source) = source
  def parse(raw) = raw.split("\n").map { |l| l.split(",") }
  def filter(rows) = rows.reject { |r| r.first.start_with?("#") }
end

if __FILE__ == $PROGRAM_NAME
  csv = "a,1\n#skip,2\nb,3"
  puts CsvImport.new.run(csv)
end

File: skills/code-architect/references/code-examples/swift/TemplateMethod.swift
// Template Method — algorithm skeleton in a base class with overridable steps.

class DataExporter {
    // The template method defines the fixed sequence of steps.
    final func export() -> String {
        let header = makeHeader()
        let body = makeBody()
        return header + "\n" + body
    }

    func makeHeader() -> String { fatalError("override makeHeader()") }
    func makeBody() -> String { fatalError("override makeBody()") }
}

final class CSVExporter: DataExporter {
    override func makeHeader() -> String { "id,name" }
    override func makeBody() -> String { "1,Ada\n2,Lin" }
}

final class MarkdownExporter: DataExporter {
    override func makeHeader() -> String { "| id | name |" }
    override func makeBody() -> String { "| 1 | Ada |\n| 2 | Lin |" }
}

func runDemo() {
    print("CSV:")
    print(CSVExporter().export())
    print("\nMarkdown:")
    print(MarkdownExporter().export())
}

runDemo()
